### ISSUE TITLE
Issues with SSL certificate validation in logstash output http plugin 

### ISSUE THREAD

**Issue by suraj-kamath** [logstash-plugins/logstash-output-http#21]
There are two issues:
1. ssl_certificate_validation option is not taking effect. Irrespective of whether we give the value as true or false, it is always taken as true.
2. During SSL certification validation, logstash is doing a strict host name check. As part of this, it is not supporting wildcard certificates where CN is like *.subdomain.domain.com. Either this needs to be fixed or it would be ideal to have another option to the http plugin like, for example, "strict_host name_check" which can be set to true or false.

Logstash version :1.5.4
Http plugin version :1.1.0
java : 1.8.0_51.

**Comment by ayashjorden** [comment]
:+1: 
Also experienced this behaviour.
@suraj-kamath nice description of the problem :)

**Comment by suraj-kamath** [comment]
@logstash-dev's can we have an update here ?

**Comment by ayashjorden** [comment]
@suraj-kamath I've found that 'ssl_certificate_validation' is used [here](https://github.com/logstash-plugins/logstash-mixin-http_client/blob/master/lib/logstash/plugin_mixins/http_client.rb).

Does it help?

**Comment by deeptjos** [comment]
Still awaiting response on this from "logstash-dev"
Please provide an update on this issue, Is there any plan of adding the changes in latest release ?

**Comment by suyograo** [comment]
> ssl_certificate_validation option is not taking effect. Irrespective of whether we give the value as true or false, it is always taken as true.

@suraj-kamath is this plugin still trying to do cert validation when `ssl_certificate_validation` is false?

**Comment by clausy** [comment]
I'm getting the same error when using
ssl_certificate_validation => false
I get 
"error" => "PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target"

I just downloaded logstash 2.0 and installed the http_poller plugin

**Comment by deeptjos** [comment]
@logstash-dev's  Any update on this ?
Also looks like Logstash does not support SNI, Please confirm.

**Comment by type0lang** [comment]
+1

**Comment by sameerpanicker** [comment]
+1

**Comment by NikolaeVarius** [comment]
+1 Getting same issue as above

**Comment by rlwmmw** [comment]
+1 Need a workaround!

**Comment by lifeofguenter** [comment]
+1

**Comment by sameerpanicker** [comment]
I was able to fix this problem. Check whether your application is using JDK or JRE. Based on that try installing the certificate in the keystore.
Let me know if you have any queries.

**Comment by dandrestor** [comment]
Same problem here.

**Comment by claudekenni** [comment]
same problem here:

"PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target"

**Comment by RohanKumbhar1** [comment]
@sameerpanicker :  great !! could you please help us here to fix this issue? may be list of steps to fix this  would  be great.

**Comment by sameerpanicker** [comment]
Go to Environment Variables and check if you are using JRE or JDK path as a JAVA_HOME value.
i.e. C:\Program Files\java\jdk1.8.0_74 or C:\Program Files\java\jre1.8.0_74

Based on that change JRE/JDK and execute this command using CMD

>C:\Program Files\Java\jre1.8.0_74\bin
>keytool -import -noprompt -trustcacerts -alias "CERT-NAME" -file "C:\Users\MYWORK\Desktop\QA.cer" -keystore "C:\Program Files\Java\jdk1.8.0_74\jre\lib\security\cacerts" -storepass changeit

Let me know if this works.

**Comment by dandrestor** [comment]
I managed to find the problem in my case, although I don't know if this would apply to everybody here.
See https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/433

**Comment by andremilk** [comment]
I am also experiencing the same issue:
https://gist.github.com/andremilk/51815410253e8b4fa86b456e254e5772

**Comment by amulyas** [comment]
+1 same issue any workaround here?

**Comment by breml** [comment]
Creating my own truststore with the `keytool` and using the configuration options `keystore` and `keystore_password` did not work for me. Adding the same certificate the same way to the `$JAVA_HOME/jre/lib/security/cacerts` truststore solved the problem for me.

Using logstash version 2.4.1.

**Comment by slalomnut** [comment]
breml's work around did the trick for me.

**Comment by bernielomax** [comment]
Yeah Im experiencing this too!

**Comment by blacklobo** [comment]
@dandrestor output HTTP does not have an valid setting of "ssl".  So that solution will not work.

This problem is still present in Logstash version 5.3.1

**Comment by yaoyaminaco0571** [comment]
@bernielomax as the error output mentioned "logstash unable to find valid certification path to requested target", you should add a fake path to it since the code needs.

The configuration is worked fine for me:
`    ssl => true
    ssl_certificate_verification => false
    truststore => "/home/admin/server/elasticsearch-current/config/truststore.jks"
    truststore_password => changeit
    user => logstash
    password => logstash`

If OK pls feedback.

**Comment by Petrox** [comment]
This issue is open since 2015.

**Comment by AverageS** [comment]
@Petrox +1 
Issue is opened since 2015 still option ssl_certificate_verification => false doesnt do anything.

**Comment by jordansissel** [comment]
I removed a few comments that referenced the Elasticsearch output. This issue and repository are for the HTTP output, not the Elasticsearch output.

**Comment by whm0627** [comment]
the issue is in the core at https://github.com/logstash-plugins/logstash-mixin-http_client
I made a pull request, but not sure when that would get merged.

for now, if you want a dirty fix, go to local file: logstash/vendor/bundle/jruby/1.9/gems/logstash-mixin-http_client-5.2.1/lib/logstash/plugin_mixins/http_client.rb
and change it to the updated one:
https://github.com/whm0627/logstash-mixin-http_client/blob/master/lib/logstash/plugin_mixins/http_client.rb

after that, set ssl_certificate_verification to false

**Comment by mytunguyen** [comment]
Hi there! I'm running logstash-5.6.3 and that "dirty fix" doesn't work either. I'm still getting `message=>"PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target"`

**Comment by whm0627** [comment]
@mytunguyen hihi. mind to post your full config? make sure there isn't any other ssl related config there.

**Comment by sureshsubbaiah** [comment]
Hi, I am running on logstash 6.0 version getting below error..
ERROR][logstash.outputs.http    ] [HTTP Output Failure] Could not fetch URL `{:url=>"https://lb:5061/evmon/tenant/instance/logging-monitoring/host/1edc775e6211?prefix=logstash-graphite.", :method=>:post, :body=>"{\"events.count\":\"204235\",\"events.rate_1m\":\"357.71489242389976\",\"events.rate_5m\":\"1794.3914592909728\",\"events.rate_15m\":\"2303.8401397055923\"}", :headers=>{"Content-Type"=>"application/json"}, :message=>"PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target", :class=>"Manticore::ClientProtocolException", :backtrace=>nil, :will_retry=>true}`

logstash http output conf
```
output {
   http
                {
                        content_type => "application/json"
                        format => "json"
                        http_method => "post"
                        url => "https://lb:5061/evmon/tenant/instance/logging-monitoring/host/1edc775e6211?prefix=logstash-graphite."
                        truststore => "/usr/share/logstash/pipeline/ssl/cacerts"
                        truststore_password => "changeit"
                        mapping => [ "events.count", "%{[events][count]}", "events.rate_1m", "%{[events][rate_1m]}", "events.rate_5m", "%{[events][rate_5m]}", "events.rate_15m", "%{[events][rate_15m]}" ]
                }
}
```

**Comment by cwstrommer** [comment]
@mytunguyen Are you sure that is from the elasticsearch output?  In my setup I found that with ```ssl_certificate_verification => false``` on the output the error came from xpack.monitoring.elasticsearch.url.   Logstash 5.6.4

**Comment by hmacias** [comment]
I worked around this issue following the instructions in https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/433

It turns out that you have to pass in **ssl => true** option in order for **ssl_certificate_verification** to take effect like so:

```
  elasticsearch {
    hosts => ["https://localhost:9200"]
    ssl => true
    ssl_certificate_verification => false
  }
```

**Comment by bnr242003** [comment]
It worked for me. Thanks a lot for your help. You saved my time.

**Comment by anottrott** [comment]
Still having this issue in 7.3.... @hmacias solution above does not work. I also tried with a known valid certificate following the instructions here, but that doesn't work either
https://www.elastic.co/guide/en/logstash/current/ls-security.html#ls-http-ssl

Is there anyone from Elastic that can comment on how to resolve this?

**Comment by normtraxler** [comment]
@anottrott : if you are using output/http (not output/elasticsearch mention by hmacias above), try my ruby code fix at https://github.com/logstash-plugins/logstash-mixin-http_client/pull/34 (file lib/logstash/plugin_mixins/http_client.rb).

**Comment by sgreszcz** [comment]
Can we get the "ignore" flag for SSL certs put back on for this output as well as inputs such as the http_poller?

```
ssl => true
ssl_certificate_verification => false
```

https://github.com/logstash-plugins/logstash-input-http_poller/issues/115

**Comment by castillar** [comment]
Can confirm this is still open. We have a Logstash server attempting to push output over HTTP to a host behind a load-balancer, with an SSL certificate that has multiple SAN entries. It looks like this:
logstash-server ---> http-receiver1.mydomain.com

The cert for "http-receiver1.mydomain.com" has "CN=someotherreceiver.mydomain.com" in the Subject, and then a SAN entry for "http-receiver1.mydomain.com". As soon as we enable the http output for this configuration, the logs fill up with this:
```
[HTTP Output Failure] Could not fetch URL {:url=>"https://http-receiver1.mydomain.com", :method=>:post, :body=>"{//SNIP LOG ENTRY HERE//}, :message=>"Host name 'http-receiver1.mydomain.com' does not match the certificate subject provided by the peer (CN=someotherreceiver.mydomain.com, O=\"Some Company, Inc.\", L=McMurdo Station, ST=NONE, C=ANTARTCTICA)", :class=>"Manticore::UnknownException", :backtrace=>["/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/manticore-0.6.4-java/lib/manticore/response.rb:37:in `block in initialize'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/manticore-0.6.4-java/lib/manticore/response.rb:79:in `call'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-http-5.2.4/lib/logstash/outputs/http.rb:239:in `send_event'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-http-5.2.4/lib/logstash/outputs/http.rb:175:in `send_events'", "/usr/share/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-output-http-5.2.4/lib/logstash/outputs/http.rb:124:in `multi_receive'", "org/logstash/config/ir/compiler/OutputStrategyExt.java:118:in `multi_receive'", "org/logstash/config/ir/compiler/AbstractOutputDelegatorExt.java:101:in `multi_receive'", "/usr/share/logstash/logstash-core/lib/logstash/java_pipeline.rb:262:in `block in start_workers'"], :will_retry=>false}
```
I've tried using the ssl_certificate_verification flag as well as the corresponding flag for the tcp output, and Logstash refuses to start with either one set, claiming it doesn't recognize those options. Could we get SAN parsing please? (Turning off SSL verification is a really bad idea.)

**Comment by b2az** [comment]
Up - i need dis damnit.

**Comment by ashmilhussain** [comment]
+1 same issue as everyone else.

**Comment by NihilBabu** [comment]
same issue need help urgently

**Comment by agberoz** [comment]
+1 same issue here

**Comment by samrui** [comment]
+1 same issure

**Comment by taajuni2** [comment]
+ 1 same issue here

**Comment by pedrolauro** [comment]
+1 Same issue here...

**Comment by fkellner** [comment]
+1 same issue with a wildcard certificate not being recognized here.

**Comment by jwhiteman36** [comment]
+1 same issue.

**Comment by Beeez** [comment]
+1 Having the same problem

**Comment by GreggBzz** [comment]
+1, same issue using wildcards and/or ignoring certification verification.

**Comment by qwerty1q2w** [comment]
+1

**Comment by mazurdv** [comment]
+1

**Comment by mldaali** [comment]
+1

**Comment by dchsueh** [comment]
+1

**Comment by denisvll** [comment]
+1

**Comment by sennl** [comment]
+1

**Comment by thunderwood19** [comment]
Since 2015...

**Comment by zgfh** [comment]
+1

**Comment by M9k** [comment]
+1

**Comment by smnschndr** [comment]
+1

**Comment by jeffwong-mocs** [comment]
@yaoyaminaco0571 's suggestion to add truststore path worked for me in v7.16.1

> @bernielomax as the error output mentioned "logstash unable to find valid certification path to requested target", you should add a fake path to it since the code needs.
> 
> The configuration is worked fine for me:
> 
> ```json
>     ssl => true
>     ssl_certificate_verification => false
>     truststore => "/home/admin/server/elasticsearch-current/config/truststore.jks"
>     truststore_password => changeit
>     user => logstash
>     password => logstash
> ```
> 
> If OK pls feedback. I guess it will work around the problem met in http output. @blacklobo
> 
> > Caution: the jks file should be exists  and readable for logstash.

**Comment by yaoyaminaco0571** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by bastianhjaeger** [comment]
Facing the same issue (out of a sudden) with no cert used at all .

This is my output section:

```
  elasticsearch {
    hosts => "https://elastic.xxx.de:9200"
    ssl => true
    ssl_certificate_verification => false
    document_id => "someid"
    user => "user"
    password => "password"
    doc_as_upsert => true
    action => "update"
  }
```

And I have no truststore to set ot a cacert.

**Comment by kares** [comment]
A lot has been going on here over the years, but let me tackle this down:
- the http_client mixin and thus output-http plugin at some point had the `ssl_certificate_validation` option
- the `ssl_certificate_validation => false` had no effect and was later removed to confuse users
- https://github.com/logstash-plugins/logstash-output-http/pull/126 added a new option `ssl_verification_mode`

The updated http output which can be used to disable verification (using `ssl_verification_mode => none`) will be part of Logstash 8.1, in the mean time try `bin/logstash-plugin update logstash-output-http` (which should update the plugin to **>= 5.3.0**)

The issue also mentions ES output with the `ssl_certificate_verification => false` option, the issue while similar do not have the same cause - for ES output disabling verification had some effect (allowing self-signed certificates) but did not disable verification completely. This [issue](https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/433) has also [been resolved](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/1063) and should be available since Logstash 7.17.0.

**Comment by yaoyaminaco0571** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
